### ISSUE TITLE
ray.init with localhost address

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#1380]
### Describe the problem

It would be nice to be able to ray.init() with a localhost address, that way your code doesn't need to change when you move ray clusters. Currently this doesn't seem to work:

```
>>> ray.init(redis_address="localhost:6379")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/anaconda3/lib/python3.6/site-packages/ray/worker.py", line 1442, in init
    redis_address = services.address_to_ip(redis_address)
  File "/home/ubuntu/anaconda3/lib/python3.6/site-packages/ray/services.py", line 187, in address_to_ip
    ip_address = socket.gethostbyname(address_parts[0])
socket.gaierror: [Errno -2] Name or service not known
```

**Comment by robertnishihara** [comment]
The following works.

```
ray start --head --node-ip-address=localhost --redis-port=6379
```

```python
import ray
ray.init(redis_address='localhost:6379')
```

Under the hood, `localhost` gets replaced with something like `127.0.0.1`. However, this won't work on multiple machines.

**Comment by ericl** [comment]
If by not working on multiple machines you mean not working on a cluster -- then that defeats the whole point :)

There are two cases I can see for ray.init:
1) Local cluster -- ray.init() with no args.
2) Actual cluster -- ray.init(redis_addr=<addr>)

In the second case, since <addr> is essentially a random value, basically you can't have scripts that work portably across multiple clusters -- you have to patch the script to point to the node runtime IP.

Actually the ideal API to just always use ray.init(), and allow switching to actual cluster mode via a env var, e.g. `RAY_CLUSTER=localhost script.py`. That way script code never needs to change.

**Comment by richardliaw** [comment]
To elaborate, I want to run the following file on any driver:

```
def experiment(): pass

if __name__ == "__main__":
    ray.init(redis_address="localhost:6379")

    tune.run_experiments({a: experiment, num_workers: 5})
```

Right now, in order to do that, I will need to modify the redis_address to include the current IP of the node I'm on. It would be great if I could didn't need to do that.

**Comment by richardliaw** [comment]
I guess a workaround would be `ray.services.get_node_ip_address() + ":6379"`

**Comment by robertnishihara** [comment]
Should be fixed by #1556.

**Comment by dHannasch** [comment]
@ericl @richardliaw Is this needed on all nodes, or only on head nodes?

I'm not sure I'm quite understanding, but it seems like if the the node were *not* the head node, then the `redis_address` (well, `address` these days https://github.com/ray-project/ray/blob/master/python/ray/worker.py#L662) would be the address of the head node, rather than localhost, right?

That is, would it be sufficient to be able to pass `localhost` and have it work (read: get replaced by `ray.services.get_node_ip_address()`) on head nodes?
